R/graded.R

```diff
@@ -1,27 +1,123 @@
+#'   
+#' @examples
+#' # Suppose our exercise asks the student to prepare and execute code that
+#' # returns the value `42`. We'll use `grade_this()` to check their
+#' # submission.
+#' #
+#' # Because we are demonstrating these functions inside R documentation, we'll
+#' # save the function returned by `grade_this()` as `this_grader()`. Calling
+#' # `this_grader()` on a mock exercise submission is equivalent to running the
+#' # check code when the student clicks "Submit Answer" in a learnr tutorial.
+#' 
+#' this_grader <- 
+#' # ```{r *-check}
+#'   grade_this({
+#'     # Automatically use .result to compare to an expected value
+#'     pass_if_equal(42, "Great work!")
+#'     
+#'     # Similarly compare .result to an expected wrong value
+#'     fail_if_equal(41, "You were so close!")
+#'     fail_if_equal(43, "Oops, a little high there!")
+#'     
+#'     # or automatically pass if .result is equal to .solution
+#'     pass_if_equal(message = "Great work!")
+#'     
+#'     # Be explicit if you need to round to avoid numerical accuracy issues
+#'     pass_if_equal(x = round(.result), y = 42, "Close enough!")
+#'     fail_if_equal(x = round(.result), y = 64, "Hmm, that's not right.")
+#'     
+#'     # For more complicated calculations, call pass() or fail()
+#'     if (.result > 100) {
+#'       fail("{.result} is way too high!")
+#'     }
+#'     if (.result * 100 == .solution) {
+#'       pass("Right answer, but {.result} is two orders of magnitude too small.")
+#'     }
+#'     
+#'     # Choose a default grade if none of the above have resulted in a grade
+#'     fail()
+#'   })
+#' # ```
+#' 
+#' # Now lets try with a few different student submissions ----
+#' 
+#' # Correct!
+#' this_grader(mock_this_exercise(.user_code = 42))
+#' 
+#' # These were close...
+#' this_grader(mock_this_exercise(.user_code = 41))
+#' this_grader(mock_this_exercise(.user_code = 43))
+#' 
+#' # Automatically use .solution if you have a *-solution chunk...
+#' this_grader(mock_this_exercise(.user_code = 42, .solution_code = 42))
+#' 
+#' # Floating point arithmetic is tricky...
+#' this_grader(mock_this_exercise(.user_code = 42.000001, .solution_code = 42))
+#' this_grader(mock_this_exercise(.user_code = 64.123456, .solution_code = 42))
+#' 
+#' # Complicated checking situations...
+#' this_grader(mock_this_exercise(.user_code = 101, .solution_code = 42))
+#' this_grader(mock_this_exercise(.user_code = 0.42, .solution_code = 42))
+#' 
+#' # Finally fall back to the final answer...
+#' this_grader(mock_this_exercise(.user_code = 33, .solution_code = 42))
 #'
 #' @param message A character string of the message to be displayed.
 #' @param correct A logical value of whether or not the checked code is correct.
-#' @describeIn graded Programatic function to produce a graded a result.
+#' @param ... Additional properties added to the graded condition. If `type`
+#'   and `location` are present, they will determine the type and location
```

Comment:
Since we are specifically talking about `type` and `location`, should we add them after the `...` with `NULL` defaults?

Reply:
That makes sense, initially I only included `...` in graded, but then I wanted `env` protected by the `...` and I decided to add the `...` to all. Now that it's there I think it makes sense to include them set to `NULL`

Reply:
If they are added, then the `...` should be blank. So maybe a `ellipsis::check_dots_empty()`